src/unity/python/turicreate/data_structures/sarray.py

```diff
@@ -349,8 +349,8 @@ def __init__(self, data=[], dtype=None, ignore_cast_failure=False, _proxy=None):
 
         if (_proxy):
             self.__proxy__ = _proxy
-        elif type(data) == SArray:
-            self.__proxy__ = data.__proxy__
+        elif isinstance(data, SArray):
+            self.__proxy__ = data.astype(dtype).__proxy__
```

Comment:
We should test if dtype is None or not.  If it is None, we should just do the old assignment, but otherwise we should call `astype`.

Reply:
If `astype(None)` is a no-op, then why not go with the simpler (no conditional) logic?

Reply:
It's not: 

```
In [2]: import turicreate as tc

In [3]: x = tc.SArray([1,2,3,4])

In [4]: x
Out[4]:
dtype: int
Rows: 4
[1, 2, 3, 4]

In [5]: y = x.astype(None)
Exception AttributeError: "'NoneType' object has no attribute '__name__'" in 'turicreate.cython.cy_flexible_type._secondary_get_translation_code' ignored
```

Thus we should add the above logic to this code, then add a test to this PR to make sure it behaves correctly when dtype is not supplied.

Reply:
Oh, I see, good call.